components/brave_shields/browser/ad_block_combinator_filters_provider.cc

```diff
@@ -0,0 +1,107 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_shields/browser/ad_block_combinator_filters_provider.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/logging.h"
+
+namespace brave_shields {
+
+AdBlockCombinatorFiltersProvider::AdBlockCombinatorFiltersProvider() {}
+
+AdBlockCombinatorFiltersProvider::~AdBlockCombinatorFiltersProvider() = default;
+
+void AdBlockCombinatorFiltersProvider::CombinateProvider(
+    AdBlockFiltersProvider* provider) {
+  auto it = std::find(combinated_providers_.begin(),
+                      combinated_providers_.end(), provider);
+  CHECK(it == combinated_providers_.end());
+  combinated_providers_.push_back(provider);
+  provider->AddObserver(this);
+}
+
+void AdBlockCombinatorFiltersProvider::UncombinateProvider(
+    AdBlockFiltersProvider* provider) {
+  auto it = std::find(combinated_providers_.begin(),
+                      combinated_providers_.end(), provider);
+  CHECK(it != combinated_providers_.end());
+  (*it)->RemoveObserver(this);
+  combinated_providers_.erase(it);
+  AdBlockFiltersProvider::OnNewSourceAvailable();
+}
+
+void AdBlockCombinatorFiltersProvider::OnNewSourceAvailable() {
```

Comment:
why are you overriding this if you're just calling the base class implementation?

Reply:
It's confusing because of the identical names. But actually the combinator implements both `AdBlockFiltersProvider` and `AdBlockFiltersProvider::Observer`; in this case the overridden method is from the `Observer`. i.e. it is forwarding the notification from _something it's observing_ to _something observing it_.

Reply:
Updated to `NotifyObservers` and `OnChanged`